### PR TITLE
Unify the addQuantDequantNode api for inputs and outputs from quant nodes

### PR DESCRIPTION
**Pull request by nishantpdce**
Stack:
&nbsp;&nbsp;&nbsp;&nbsp;:black_circle:&nbsp; **#20677 Unify the addQuantDequantNode api for inputs and outputs from quant nodes**&nbsp;&nbsp;[:yellow_heart:](https://our.internmc.facebook.com/intern/diff/D15406354/)

With new changes in IR, it is possible to insert nodes after param
nodes in graph. Thus we do not need to have two methods for inserting q-dq
nodes to input or output to quantizable nodes.

Differential Revision: [D15406354](https://our.internmc.facebook.com/intern/diff/D15406354/)

### CODE REVIEW

File: torch/csrc/jit/passes/quantization.cpp [hunk 1]
```diff
@@ -463,17 +426,26 @@ void InsertQuantDequantNodes(
...
-      addQuantDeQuantNodes(ele, qparam_value_dict[ele]);
+  for (auto& v_to_quant : quantOutputs) {
+    if (qparam_value_dict.count(v_to_quant) != 0) {
+      Node* dq = addQuantDeQuantNodesFor(
+          v_to_quant, v_to_quant->node(), qparam_value_dict[v_to_quant]);
+      TORCH_INTERNAL_ASSERT(dq != nullptr);
+      v_to_quant->replaceAllUsesWith(dq->output());
+      // Above step replaces v->quant with vdq->quant. We need to restore link.
+      // Below chain traverse up from dq to q node.
+      dq->inputs()[0]->node()->inputs()[0]->node()->replaceInputWith(
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Maybe factor `dq->inputs()[0]->node()->inputs()[0]->node()` out to a separate function and add asserts there?

File: torch/csrc/jit/passes/quantization.cpp [hunk 2]
```diff
@@ -373,7 +336,7 @@ void InsertQuantDequantNodes(
   blocks_to_visit.push(graph->block());
   // For storing quantizable values - node pairs that are external
   // or intermediate inputs to quantizable nodes
-  std::vector<std::pair<Value*, Node*>> quantInputs;
+  std::vector<param_info_t> quantInputs;
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: This change seems unrelated.

File: torch/csrc/jit/passes/quantization.cpp [hunk 3]
```diff
@@ -144,21 +144,21 @@ Node* createIntReprNode(Value* v, Node* n) {
...
+    Node* insert_point,
     const std::tuple<std::string, float, int>& qparam,
     at::ScalarType t = at::ScalarType::Undefined) {
   TORCH_INTERNAL_ASSERT(v != nullptr);
-  Node* n = v->node();
-  Node* quant = createQuantNode(v, n);
-  Node* intrepr = createIntReprNode(v, n);
-  Node* dequant = createDeQuantNode(v, n);
+  WithInsertPoint ins(insert_point);
+  Node* quant = createQuantNode(v, insert_point);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: How about inlining these calls here? There is nothing much happening inside anyway - asserts are already there and the scope stuff can be removed if we use `WithCurrentScope` (it works similar to `WithInsertPoint` but sets the scope rather than the insert point).

If we do this and hoist `WithInsertPoint ins` out of this function to its callsite we would be able to get rid of the second argument (`Node* insert_point`).

File: torch/csrc/jit/passes/quantization.cpp [hunk 4]
```diff
@@ -144,21 +144,21 @@ Node* createIntReprNode(Value* v, Node* n) {
...
+  Node* intrepr = createIntReprNode(v, insert_point);
+  Node* dequant = createDeQuantNode(v, insert_point);
 
   // Add quant-intrepr-dequant nodes and replace for all uses of Value
-  quant->insertAfter(n);
+  quant->insertAfter(insert_point);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Nit: since you've already set up the insert point, you could use `insert` instead of `insertAfter`. That would actually help to make sure we're inserting everything in the same place.

**Review by ZolotukhinM — COMMENTED** [review]
Cool, I think it moves in the right direction - very nice to see diffs removing code without losing any functionality and improving the readability of the codebase.

I also see some ways to clean it up even more, please take a look inline!

File: torch/csrc/jit/passes/quantization.cpp (hunk 2 above)
Comment (nishantpdce) [reply torch/csrc/jit/passes/quantization.cpp]: This is one of ur older comment to use struct instead of pair . Its used below since we have struct defined already

File: torch/csrc/jit/passes/quantization.cpp (hunk 3 above)
Comment (nishantpdce) [reply torch/csrc/jit/passes/quantization.cpp]: I recollect one more point. we have perchannelquantization coming up, so createQuantNode() will actually get expanded to take more argument and create node based on qscheme. i think it will be beneficial to retain these api as is instead of inlining

File: torch/csrc/jit/passes/quantization.cpp (hunk 4 above)
Comment (nishantpdce) [reply torch/csrc/jit/passes/quantization.cpp]: btw are u referring to below api.  I am trying to look for usage for below api in code.  will setting the insert_point not guarantee all nodes inserted at same point ? 
TORCH_API Value* insert(
      Symbol opname,
      at::ArrayRef<NamedValue> args,
      at::ArrayRef<NamedValue> kwargs = {},
      const c10::optional<SourceRange>& range = {});

File: torch/csrc/jit/passes/quantization.cpp (hunk 1 above)
Comment (nishantpdce) [reply torch/csrc/jit/passes/quantization.cpp]: Done

File: torch/csrc/jit/passes/quantization.cpp (hunk 3 above)
Comment (nishantpdce) [reply torch/csrc/jit/passes/quantization.cpp]: Added the WithCurrentScope as suggested but retaining the createQuantNode() api due to reason stated above

File: torch/csrc/jit/passes/quantization.cpp (hunk 4 above)
Comment (nishantpdce) [reply torch/csrc/jit/passes/quantization.cpp]: i have retained same api for same reason. insert requires passing explicit opname which we want to avoid as quant node may refer to quantize_linear or quantize_linear_affine in coming future

File: torch/csrc/jit/passes/quantization.cpp [hunk 5]
```diff
@@ -30,9 +30,20 @@ int groups, bool benchmark, bool deterministic, bool cudnn_enabled) -> Tensor"};
   return quantnodeLookup.find(n);
 }
 
+void travAndReplaceQuantnodeinput(Node* dq, Value* v_to_quant) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Best practices are for a function to do only one thing - is there a reason to include `replaceInputWith` call here? That's the only reason we have the second argument here and it's completely unnecessary. Basically, we had
```
doFirstThing(a);
doSecondThing(b);
```
And this changed it to `doFirstAndSecondThing(a, b)`.

Sorry if I was not completely clear in my initial comment, but the idea was to only factor out
`dq->inputs()[0]->node()->inputs()[0]->node()`.

File: torch/csrc/jit/passes/quantization.cpp (hunk 3 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/passes/quantization.cpp]: My concern about the current implementation is that it's misleading. For instance, a call `createQuantNode(v, insert_point)` looks like it would insert whatever it does into `insert_point`, while in fact it uses this argument for just retrieving the graph from it. My previous suggestions were one potential way to resolve it. You points are also valid, so I don't object to having these functions separate, but we still need to make them well-defined and easily understandable one way or another before it gets in.

File: torch/csrc/jit/passes/quantization.cpp (hunk 4 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/passes/quantization.cpp]: There are still ways to pull it off. One potential way would be to figure out the symbol basing on the qconfig/qparams and then passing it to insert - I don't see why it would be a problem. What concerns me is that we are currently using two ways for nodes insertion - using only one would make the code more robust for potential future changes.

File: torch/csrc/jit/passes/quantization.cpp (hunk 2 above)
Comment (ZolotukhinM) [reply torch/csrc/jit/passes/quantization.cpp]: I'd rather prefer it to be addressed in the PR where I commented and not here, but that's also fine :)

File: torch/csrc/jit/passes/quantization.cpp [hunk 6]
```diff
@@ -463,17 +436,25 @@ void InsertQuantDequantNodes(
...
-      addQuantDeQuantNodes(ele, qparam_value_dict[ele]);
+  for (auto& v_to_quant : quantOutputs) {
+    if (qparam_value_dict.count(v_to_quant) != 0) {
+      Node* dq = addQuantDeQuantNodesFor(
+          v_to_quant, v_to_quant->node(), qparam_value_dict[v_to_quant]);
+      TORCH_INTERNAL_ASSERT(dq != nullptr);
+      v_to_quant->replaceAllUsesWith(dq->output());
+      // Above step replaces v->quant with vdq->quant. We need to restore link.
+      // Below chain traverse up from dq to q node.
+      travAndReplaceQuantnodeinput(dq, v_to_quant);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/quantization.cpp]: Nit1: What is `trav`? Is it `traverse`?
Nit2: Capitalize first letters in **n**ode**i**nput in the function name.

**Review by ZolotukhinM — CHANGES_REQUESTED** [review]
I still see some issues outstanding, please take a look at my replies.

File: torch/csrc/jit/passes/quantization.cpp (hunk 4 above)
Comment (nishantpdce) [reply torch/csrc/jit/passes/quantization.cpp]: sorry i did not follow what are the two ways :) for node insertion. I also wanted to match the mechanism we use in addObserver function that u added. There we use insertAfter instead of insert. I think if we are to use insert, it should be same everywhere - agree

File: torch/csrc/jit/passes/quantization.cpp (hunk 3 above)
Comment (nishantpdce) [reply torch/csrc/jit/passes/quantization.cpp]: I can add a note and think of better name. if u have something in mind please suggest

File: torch/csrc/jit/passes/quantization.cpp (hunk 6 above)
Comment (nishantpdce) [reply torch/csrc/jit/passes/quantization.cpp]: i will make this change all together once i hear for your response on above

File: torch/csrc/jit/passes/quantization.cpp (hunk 5 above)
Comment (nishantpdce) [reply torch/csrc/jit/passes/quantization.cpp]: i will make this change all together once i hear for your response on below items

File: torch/csrc/jit/passes/quantization.cpp (hunk 6 above)
Comment (nishantpdce) [reply torch/csrc/jit/passes/quantization.cpp]: Done

File: torch/csrc/jit/passes/quantization.cpp (hunk 3 above)
Comment (nishantpdce) [reply torch/csrc/jit/passes/quantization.cpp]: Done based on above comment

File: torch/csrc/jit/passes/quantization.cpp (hunk 5 above)
Comment (nishantpdce) [reply torch/csrc/jit/passes/quantization.cpp]: Done

**Review by Krovatkin — APPROVED** [review]
:shipit:

File: torch/csrc/jit/passes/quantization.cpp (hunk 4 above)
Comment (Krovatkin) [reply torch/csrc/jit/passes/quantization.cpp]: since it's only a nitpick thing according to @ZolotukhinM , let's just remove `WithInsertionPoint` and keep `insertAfter` for now.

File: torch/csrc/jit/passes/quantization.cpp (hunk 2 above)
Comment (Krovatkin) [reply torch/csrc/jit/passes/quantization.cpp]: i'm also fine with this change in this PR.

**Review by Krovatkin — APPROVED** [review]
